communication/src/TlsTcpImpl.h

```diff
@@ -0,0 +1,252 @@
+  bool isRunning() const;
+  ConnectionStatus getCurrentConnectionStatus(const NodeNum node) const;
+  void setReceiver(NodeNum nodeId, IReceiver *receiver);
+  int sendAsyncMessage(const NodeNum destination, const char *const msg, const size_t msg_len);
+  int getMaxMessageSize();
+
+ private:
+  // Open a socket, configure it, bind it, and start listening on a given host/pot.
+  void listen();
+
+  // Start asynchronously accepting connections.
+  void accept();
+
+  // Perform synhronous DNS resolution. This really only works for the listen socket, since after that we want to do a
+  // new lookup on every connect operation in case the underlying IP of the DNS address changes.
+  //
+  // Throws a boost::system::system_error if it fails to resolve
+  boost::asio::ip::tcp::endpoint syncResolve();
+
+  // Starts async dns resolution on a tcp socket.
+  //
+  // Every call to connect will call resolve before asio async_connect call.
+  void resolve(NodeNum destination);
+
+  // Connet to other nodes where there is not a current connection.
+  //
+  // Replicas only connect to replicas with smaller node ids.
+  // This method is called at startup and also on a periodic timer tick.
+  void connect();
+  void connect(NodeNum, boost::asio::ip::tcp::endpoint);
+
+  // Start a steady_timer in order to trigger needed `connect` operations.
+  void startConnectTimer();
+
+  // Trigger the asio async_hanshake calls.
+  void startServerSSLHandshake(boost::asio::ip::tcp::socket &&);
+  void startClientSSLHandshake(boost::asio::ip::tcp::socket &&socket, NodeNum destination);
+
+  // Callbacks triggered when asio async_hanshake completes for an incoming or outgoing connection.
+  void onServerHandshakeComplete(const boost::system::error_code &ec, size_t accepted_connection_id);
+  void onClientHandshakeComplete(const boost::system::error_code &ec, NodeNum destination);
+
+  // If onServerHandshake completed successfully, this function will get called and add the AsyncTlsConnection to
+  // `connections_`.
+  void onConnectionAuthenticated(std::shared_ptr<AsyncTlsConnection> conn);
+
+  // Synchronously close connections
+  // This is only used during shutdown after the io_service is stopped.
+  void syncCloseConnection(std::shared_ptr<AsyncTlsConnection> &conn);
+
+  // Asynchronously shutdown an SSL connection and then close the underlying TCP socket when the shutdown has
+  // completed.
+  void closeConnection(NodeNum);
+  void closeConnection(std::shared_ptr<AsyncTlsConnection> conn);
+
+  bool isReplica() { return config_.selfId <= static_cast<size_t>(config_.maxServerId); }
+
+  logging::Logger logger_;
+  TlsTcpConfig config_;
+
+  // The lifetime of the receiver must be at least as long as the lifetime of this object
+  IReceiver *receiver_;
+
+  // This thread runs the asio io_service. It's the main thread of the TlsTcpImpl.
+  std::unique_ptr<std::thread> io_thread_;
+
+  // io_thread_ lifecycle management
+  mutable std::mutex startStopGuard_;
```

Comment:
We should rename this to `io_thread_guard_`

Reply:
Yup. I also changed `connectionsGuard_` to `connections_guard_`.